services/clusters-keeper/src/simcore_service_clusters_keeper/modules/clusters_management_core.py

```diff
@@ -0,0 +1,79 @@
+import datetime
+import logging
+
+import arrow
+from fastapi import FastAPI
+
+from ..core.settings import get_application_settings
+from ..models import EC2InstanceData
+from ..modules.clusters import delete_clusters, get_all_clusters, set_instance_heartbeat
+from ..utils.dask import get_gateway_authentication, get_gateway_url
+from ..utils.ec2 import HEARTBEAT_TAG_KEY, get_user_id_from_tags
+from .dask import is_gateway_busy, ping_gateway
+
+_logger = logging.getLogger(__name__)
+
+
+def _get_instance_last_heartbeat(instance: EC2InstanceData) -> datetime.datetime:
+    if last_heartbeat := instance.tags.get(HEARTBEAT_TAG_KEY, None):
+        last_heartbeat_time: datetime.datetime = arrow.get(last_heartbeat).datetime
+        return last_heartbeat_time
+    return instance.launch_time
+
+
+async def _find_terminateable_instances(
```

Comment:
terminator instances? :-)

Reply:
ask AWS... that is not my nomenclature...

Reply:
ask AWS... that is not my nomenclature...